components/content/content-editor/content-editor.service.ts

```
@@ -0,0 +1,185 @@
+			return null;
+		}
+		return node;
+	}
+
+	private static findParentNode(proposedParent: Node, child: Node): Node | null {
+		for (let i = 0; i < proposedParent.childCount; i++) {
+			const parentChild = proposedParent.child(i);
+			if (parentChild === child) {
+				return proposedParent;
+			} else {
+				const maybeParent = this.findParentNode(parentChild, child);
+				if (maybeParent !== null) {
+					return maybeParent;
+				}
+			}
+		}
+
+		return null;
+	}
+
+	public static getParentNode(state: EditorState, child: Node) {
+		return this.findParentNode(state.doc, child);
+	}
+
+	/**
+	 * Returns a list of all applied to any node within the selection
+	 */
+	public static getSelectionMarks(state: EditorState) {
+		const markTypes: Mark[] = [];
+		state.doc.nodesBetween(
+			state.selection.from,
+			state.selection.to,
+			(node: Node, _0: number, _1: Node, _2: number) => {
+				for (const mark of node.marks) {
+					if (!markTypes.some(m => m.type.name === mark.type.name)) {
+						markTypes.push(mark);
+					}
+				}
+			}
+		);
+		return markTypes;
+	}
+
+	/**
+	 * Indicates whether the given node is contained in a node of the given type.
+	 * Returns the parent node that matched, or `false`.
+	 */
+	public static isContainedInNode(state: EditorState, node: Node, nodeType: NodeType) {
+		if (node.type.name === nodeType.name) {
+			return node;
+		}
+
+		let child = node;
+		let parent = this.getParentNode(state, node);
+		while (parent !== null) {
+			if (parent.type.name === nodeType.name) {
+				return parent;
+			}
+
+			// Walk up
+			child = parent;
+			parent = this.getParentNode(state, child);
+		}
+		return false;
+	}
+
+	public static handleImageUploads(view: EditorView, items: DataTransferItemList) {
+		let handled = false;
+
+		for (let i = 0; i < items.length; i++) {
+			const transferItem = items[i];
+
+			if (
+				transferItem.kind === 'file' &&
+				imageMimeTypes.includes(transferItem.type.toLowerCase())
+			) {
```

Shouldn't we be throwing in this function instead of continuing?
Theres no way to tell which file was handled if multiple were passed in.

---

Also, instead of checking like this - use `transferItem.getAsFile()` and then use the isImage `from utils/image.ts`

---

getAsFile might come at a performance cost, I think this is reasonable code duplication, since isImage only checks the imageMimeTypes array, which we import from there.